loader/loader_linux.c

```diff
@@ -0,0 +1,458 @@
+VkResult linux_read_sorted_physical_devices(struct loader_instance *inst, uint32_t icd_count,
+                                            struct loader_phys_dev_per_icd *icd_devices,
+                                            struct loader_physical_device_term **sorted_device_term) {
+    VkResult res = VK_SUCCESS;
+    bool is_vulkan_1_1 = false;
+    if (inst->app_api_major_version >= 1 && inst->app_api_minor_version >= 1) {
+        is_vulkan_1_1 = true;
+    }
+
+    struct LinuxSortedDeviceInfo *sorted_device_info = loader_instance_heap_alloc(
+        inst, inst->total_gpu_count * sizeof(struct LinuxSortedDeviceInfo), VK_SYSTEM_ALLOCATION_SCOPE_COMMAND);
+    if (NULL == sorted_device_info) {
+        return VK_ERROR_OUT_OF_HOST_MEMORY;
+    }
+    memset(sorted_device_info, 0, inst->total_gpu_count * sizeof(struct LinuxSortedDeviceInfo));
+
+    loader_log(inst, VULKAN_LOADER_INFO_BIT | VULKAN_LOADER_DRIVER_BIT, 0, "linux_read_sorted_physical_devices:  Original order:");
+
+    // Grab all the necessary info we can about each device
+    uint32_t index = 0;
+    for (uint32_t icd_idx = 0; icd_idx < icd_count; ++icd_idx) {
+        for (uint32_t phys_dev = 0; phys_dev < icd_devices[icd_idx].count; ++phys_dev) {
+            struct loader_icd_term *icd_term = icd_devices[icd_idx].this_icd_term;
+            VkPhysicalDeviceProperties dev_props = {};
+
+            sorted_device_info[index].physical_device = icd_devices[icd_idx].phys_devs[phys_dev];
+            sorted_device_info[index].icd_index = icd_idx;
+            sorted_device_info[index].icd_term = icd_term;
+            sorted_device_info[index].has_pci_bus_info = false;
+
+            icd_term->dispatch.GetPhysicalDeviceProperties(sorted_device_info[index].physical_device, &dev_props);
+            sorted_device_info[index].device_type = dev_props.deviceType;
+            strncpy(sorted_device_info[index].device_name, dev_props.deviceName, VK_MAX_PHYSICAL_DEVICE_NAME_SIZE);
+            sorted_device_info[index].vendor_id = dev_props.vendorID;
+            sorted_device_info[index].device_id = dev_props.deviceID;
+
+            bool device_is_1_1_capable =
+                VK_API_VERSION_MAJOR(dev_props.apiVersion) == 1 && VK_API_VERSION_MINOR(dev_props.apiVersion) >= 1;
+            sorted_device_info[index].has_pci_bus_info = device_is_1_1_capable;
+            if (!sorted_device_info[index].has_pci_bus_info) {
+                uint32_t ext_count;
+                icd_term->dispatch.EnumerateDeviceExtensionProperties(sorted_device_info[index].physical_device, NULL, &ext_count,
+                                                                      NULL);
+                if (ext_count > 0) {
+                    VkExtensionProperties *ext_props =
+                        (VkExtensionProperties *)loader_stack_alloc(sizeof(VkExtensionProperties) * ext_count);
+                    if (NULL == ext_props) {
+                        return VK_ERROR_OUT_OF_HOST_MEMORY;
```

Comment:
Feel free to ignore, but this function is big enough it might be good to have a `goto out;` Since there are several places this could return from, and making sure no memory is leaked by accident is a good thing.

Reply:
Good catch.  I think I'll make the single exit just to make sure I don't miss any others.